// @trezor/blockchain-link/lib/workers/blockbook/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var bignumber_js_1 = __importDefault(require("bignumber.js"));
exports.transformServerInfo = function (payload) {
    return {
        name: payload.name,
        shortcut: payload.shortcut,
        testnet: payload.testnet,
        version: payload.version,
        decimals: payload.decimals,
        blockHeight: payload.bestHeight,
        blockHash: payload.bestHash,
    };
};
exports.filterTargets = function (addresses, targets) {
    if (typeof addresses === 'string') {
        addresses = [addresses];
    }
    if (!addresses || !Array.isArray(addresses) || !targets || !Array.isArray(targets))
        return [];
    var all = addresses.map(function (a) {
        if (typeof a === 'string')
            return a;
        if (typeof a === 'object' && typeof a.address === 'string')
            return a.address;
        return null;
    });
    return targets.filter(function (t) {
        if (t && Array.isArray(t.addresses)) {
            return t.addresses.filter(function (a) { return all.indexOf(a) >= 0; }).length > 0;
        }
        return false;
    });
};
exports.filterTokenTransfers = function (addresses, transfers) {
    if (typeof addresses === 'string') {
        addresses = [addresses];
    }
    if (!addresses || !Array.isArray(addresses) || !transfers || !Array.isArray(transfers))
        return [];
    var all = addresses.map(function (a) {
        if (typeof a === 'string')
            return a;
        if (typeof a === 'object' && typeof a.address === 'string')
            return a.address;
        return null;
    });
    return transfers
        .filter(function (tr) {
        if (tr && typeof tr === 'object') {
            return (tr.from && all.indexOf(tr.from) >= 0) || (tr.to && all.indexOf(tr.to) >= 0);
        }
        return false;
    })
        .map(function (tr) {
        var incoming = tr.from && all.indexOf(tr.from) >= 0;
        var outgoing = tr.to && all.indexOf(tr.to) >= 0;
        var type = 'unknown';
        if (incoming && outgoing) {
            type = 'self';
        }
        else {
            if (incoming) {
                type = 'sent';
            }
            if (outgoing) {
                type = 'recv';
            }
        }
        return {
            type: type,
            name: tr.name,
            symbol: tr.symbol,
            address: tr.token,
            decimals: tr.decimals || 0,
            amount: tr.value,
            from: tr.from,
            to: tr.to,
        };
    });
};
var transformTarget = function (target) {
    return {
        addresses: target.addresses,
        isAddress: target.isAddress,
        amount: target.value,
        coinbase: target.coinbase,
    };
};
exports.transformTransaction = function (descriptor, addresses, tx) {
    var myAddresses = addresses
        ? addresses.change.concat(addresses.used, addresses.unused)
        : [descriptor];
    var vinLength = Array.isArray(tx.vin) ? tx.vin.length : 0;
    var voutLength = Array.isArray(tx.vout) ? tx.vout.length : 0;
    var outgoing = exports.filterTargets(myAddresses, tx.vin);
    var incoming = exports.filterTargets(myAddresses, tx.vout);
    var internal = addresses ? exports.filterTargets(addresses.change, tx.vout) : [];
    var tokens = exports.filterTokenTransfers(myAddresses, tx.tokenTransfers);
    var type;
    var targets = [];
    var amount = tx.value;
    if (outgoing.length === 0 && incoming.length === 0 && tokens.length === 0) {
        type = 'unknown';
    }
    else if (vinLength > 0 &&
        voutLength > 0 &&
        outgoing.length === vinLength &&
        incoming.length === voutLength) {
        type = 'self';
        amount = tx.fees;
    }
    else if (outgoing.length === 0 && (incoming.length > 0 || tokens.length > 0)) {
        type = 'recv';
        amount = '0';
        if (incoming.length > 0) {
            if (Array.isArray(tx.vin)) {
                targets = tx.vin;
            }
            amount = incoming.reduce(function (prev, vout) {
                if (typeof vout.value !== 'string')
                    return prev;
                var bn = new bignumber_js_1.default(prev).plus(vout.value);
                return bn.toString();
            }, amount);
        }
    }
    else {
        type = 'sent';
        if (tokens.length === 0 && Array.isArray(tx.vout)) {
            targets = tx.vout.filter(function (o) { return incoming.indexOf(o) < 0 && internal.indexOf(o) < 0; });
        }
        if (tx.ethereumSpecific) {
            amount = tokens.length > 0 || tx.ethereumSpecific.status === 0 ? tx.fees : tx.value;
        }
        else if (Array.isArray(tx.vout)) {
            var myInputsSum = outgoing.reduce(function (prev, vin) {
                if (typeof vin.value !== 'string')
                    return prev;
                var bn = new bignumber_js_1.default(prev).plus(vin.value);
                return bn.toString();
            }, '0');
            amount = incoming.reduce(function (prev, vout) {
                if (typeof vout.value !== 'string')
                    return prev;
                var bn = new bignumber_js_1.default(prev).minus(vout.value);
                return bn.toString();
            }, myInputsSum);
        }
    }
    var rbf;
    if (Array.isArray(tx.vin)) {
        tx.vin.forEach(function (vin) {
            if (typeof vin.sequence === 'number' && vin.sequence < 0xffffffff - 1) {
                rbf = true;
            }
        });
    }
    return {
        type: type,
        txid: tx.txid,
        blockTime: tx.blockTime,
        blockHeight: tx.blockHeight,
        blockHash: tx.blockHash,
        amount: amount,
        fee: tx.fees,
        targets: targets.filter(function (t) { return typeof t === 'object'; }).map(function (t) { return transformTarget(t); }),
        tokens: tokens,
        rbf: rbf,
        ethereumSpecific: tx.ethereumSpecific,
    };
};
exports.transformTokenInfo = function (tokens) {
    if (!tokens || !Array.isArray(tokens))
        return undefined;
    var info = tokens.reduce(function (arr, t) {
        if (t.type !== 'ERC20')
            return arr;
        return arr.concat([
            {
                type: t.type,
                name: t.name,
                symbol: t.symbol,
                address: t.contract,
                balance: t.balance,
                decimals: t.decimals || 0,
            },
        ]);
    }, []);
    return info.length > 0 ? info : undefined;
};
exports.transformAddresses = function (tokens) {
    if (!tokens || !Array.isArray(tokens))
        return undefined;
    var addresses = tokens.reduce(function (arr, t) {
        if (t.type !== 'XPUBAddress')
            return arr;
        return arr.concat([
            {
                address: t.name,
                path: t.path,
                transfers: t.transfers,
                balance: t.balance,
                sent: t.totalSent,
                received: t.totalReceived,
            },
        ]);
    }, []);
    if (addresses.length < 1)
        return undefined;
    var internal = addresses.filter(function (a) { return a.path.split('/')[4] === '1'; });
    var external = addresses.filter(function (a) { return internal.indexOf(a) < 0; });
    return {
        change: internal,
        used: external.filter(function (a) { return a.transfers > 0; }),
        unused: external.filter(function (a) { return a.transfers === 0; }),
    };
};
exports.transformAccountInfo = function (payload) {
    var page;
    if (typeof payload.page === 'number') {
        page = {
            index: payload.page,
            size: payload.itemsOnPage,
            total: payload.totalPages,
        };
    }
    var misc;
    if (typeof payload.nonce === 'string') {
        misc = { nonce: payload.nonce };
    }
    var descriptor = payload.address;
    var addresses = exports.transformAddresses(payload.tokens);
    var tokens = exports.transformTokenInfo(payload.tokens);
    var empty = payload.txs === 0 && payload.unconfirmedTxs === 0;
    var unconfirmed = new bignumber_js_1.default(payload.unconfirmedBalance);
    var availableBalance = !unconfirmed.isNaN() && !unconfirmed.isZero()
        ? unconfirmed.plus(payload.balance).toString()
        : payload.balance;
    return {
        descriptor: descriptor,
        balance: payload.balance,
        availableBalance: availableBalance,
        empty: empty,
        tokens: tokens,
        addresses: addresses,
        history: {
            total: payload.txs,
            tokens: typeof payload.nonTokenTxs === 'number'
                ? payload.txs - payload.nonTokenTxs
                : undefined,
            unconfirmed: payload.unconfirmedTxs,
            transactions: payload.transactions
                ? payload.transactions.map(function (t) { return exports.transformTransaction(descriptor, addresses, t); })
                : undefined,
        },
        misc: misc,
        page: page,
    };
};
exports.transformAccountUtxo = function (payload) {
    return payload.map(function (utxo) { return ({
        txid: utxo.txid,
        vout: utxo.vout,
        amount: utxo.value,
        blockHeight: utxo.height,
        address: utxo.address,
        path: utxo.path,
        confirmations: utxo.confirmations,
        coinbase: utxo.coinbase,
    }); });
};
